src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2083,6 +2084,70 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
   return this;
 }
 
+Node* SelectFromTwoVectorNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  int num_elem = vect_type()->length();
+  BasicType elem_bt = vect_type()->element_basic_type();
+
+  // Keep the node if it is supported, else lower it to other nodes.
+  if (Matcher::match_rule_supported_vector(Op_SelectFromTwoVector, num_elem, elem_bt)) {
+    return nullptr;
+  }
+
+  Node* index_vec = in(1);
+  Node* src1 = in(2);
+  Node* src2 = in(3);
+
+  // Lower the IR to constituents operations.
+  //   SelectFromTwoVectorNode =
+  //     (VectorBlend
+  //         (VectorRearrange SRC1 (WRAPED_INDEX AND (VLEN-1))
+  //         (VectorRearrange SRC2 (WRAPED_INDEX AND (VLEN-1))
+  //         MASK)
+  // Where
+  //   incoming WRAPED_INDEX is within two vector index range [0, VLEN*2) and
+  //   MASK = WRAPED_INDEX < VLEN
+  //
+  // IR lowering prevents intrinsification failure and associated argument
+  // boxing penalties.
+  //
+
+  const TypeVect* index_vect_type = index_vec->bottom_type()->is_vect();
+  BasicType index_elem_bt = index_vect_type->element_basic_type();
+  assert(!is_floating_point_type(index_elem_bt), "");
+
+  // Downcast index vector to a type agnostic shuffle representation, shuffle indices
+  // are held in a byte vector which are later massaged to target specific permutation
+  // index format by subsequent VectorLoadShuffle.
+  int cast_vopc = VectorCastNode::opcode(0, index_elem_bt, true);
+  Node* index_byte_vec = phase->transform(VectorCastNode::make(cast_vopc, index_vec, T_BYTE, num_elem));
```

Comment:
This cast assumes that the indices cannot have more than 8 bits. This would allow vector lengths of up to 256. This is fine for intel. But as far as I know ARM has in principle longer vectors - up to 2048 bytes. Should we maybe add some assert here to make sure we never badly truncate the index?

Reply:
Shuffle overall is on our todo list, its a know limitation which we tried lifting once, yes you read it correctly, its a limitation for AARCH64 SVE once a 2048 bits vector systems are available, IIRC current max vector size on any available AARCH64 system is 256 bits, with Neoverse V2 they shrink the vector size back to 16 bytes.

Reply:
Are there any asserts that would catch this?